changes/test_views.py

```diff
@@ -426,3 +432,84 @@ def test_whatsapp_contact_check_failure(self, task):
 
         self.assertEqual(response.status_code, status.HTTP_202_ACCEPTED)
         task.delay.assert_called_once_with(str(user.pk), "+27820001001")
+
+
+class ReceiveSeedMessageSenderFailedMsisdnViewTests(APITestCase):
+    def make_source_normaluser(self):
+        data = {
+            "name": "test",
+            "authority": "patient",
+            "user": User.objects.get(username="test"),
+        }
+        return Source.objects.create(**data)
+
+    def subscription_lookup(self):
+        responses.add(
+            responses.GET, "http://sbm/api/v1/subscriptions/", json={}, status=200
+        )
+
+    def identity_lookup(self, identity):
+        responses.add(
+            responses.GET,
+            "http://is/api/v1/identities/%s/" % (identity),
+            json={},
+            status=200,
+        )
+
+    def jembi_post(self):
+        responses.add(
+            responses.POST, "http://jembi/ws/rest/v1/optout", json={}, status=200
+        )
+
+    def test_token_querystring_auth(self):
+        """
+        The token should be required in the querystring for requests to be
+        processed.
+        """
+        url = reverse("message_sender_failed_msisdn_hook")
+        response = self.client.post(url, data={}, format="json")
+        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
+
+    @responses.activate
+    def test_failed_msidn_lookup_hook(self):
+        """
+        If the message sender sends us an event for failed msisdn lookup
+        the hook should be called
+        """
+        user = User.objects.create_user("test")
+        token = Token.objects.create(user=user).key
+        url = "{}?{}".format(
+            reverse("message_sender_failed_msisdn_hook"), urlencode({"token": token})
+        )
+        registrant_id = "cbaf27bc-2ba9-4e4a-84c1-098d5abd80bf"
+        source = self.make_source_normaluser()
+
+        response = self.client.post(
+            url,
+            data={
+                "hook": {
+                    "id": 1,
+                    "event": "identity.no_address",
+                    "target": "http://example.org",
+                },
+                "data": {"registrant_id": registrant_id},
+            },
+            format="json",
+        )
+
+        self.subscription_lookup()
+        self.identity_lookup(registrant_id)
+        self.jembi_post()
+
+        change_data = {
+            "registrant_id": "cbaf27bc-2ba9-4e4a-84c1-098d5abd80bf",
+            "action": "momconnect_nonloss_optout",
+            "data": {"reason": "missing_to_addr"},
+            "source": source,
+        }
+        change = Change.objects.create(**change_data)
```

Comment:
We don't want to be creating the Change here in the test. The view should be creating the `Change` for us, so the `self.client.post` should create the `Change`. What we should be doing here is fetching the `Change` from the database, and asserting that the fields were created correctly.

Reply:
how do we fetch the change from the database?